builder/remotecontext/git/gitutils.go

```diff
@@ -98,22 +97,45 @@ func getRefAndSubdir(fragment string) (ref string, subdir string) {
 
 func fetchArgs(remoteURL string, ref string) []string {
 	args := []string{"fetch", "--recurse-submodules=yes"}
-	shallow := true
 
-	if urlutil.IsURL(remoteURL) {
-		res, err := http.Head(fmt.Sprintf("%s/info/refs?service=git-upload-pack", remoteURL))
-		if err != nil || res.Header.Get("Content-Type") != "application/x-git-upload-pack-advertisement" {
-			shallow = false
-		}
-	}
-
-	if shallow {
+	if supportsShallowCheckout(remoteURL) {
 		args = append(args, "--depth", "1")
 	}
 
 	return append(args, "origin", ref)
 }
 
+func supportsShallowCheckout(remoteURL string) bool {
+	// Check if the URL is an HTTP url
+	if urlutil.IsURL(remoteURL) {
+		// Only smart-HTTP git servers support shallow clone
+
+		// URL to test for the service
+		serviceURL := remoteURL + "/info/refs?service=git-upload-pack"
+		// Desired content type returned by smart servers
+		contentTypeResponse := "application/x-git-upload-pack-advertisement"
+
+		// First try a HEAD request
+		res, err := http.Head(serviceURL)
```

Comment:
I think you tried to reduce complexity here by keeping the `HEAD` and `GET` parts separate, however doing so probably made it harder to grasp;

- the `contentTypeResponse` variable is defined a lot further from where it's used
- explaining things like "if x succeeds", "check content type" should not be needed; the code can be self-explanatory

If you combine the paths, a single explanation of what we're doing can be sufficient;

```go
// try HEAD, fallback to GET if not supported
res, err := http.Head(serviceURL)
if err != nil {
	res, err = http.Get(serviceURL)
	if err != nil {
		return false
	}
	res.Body.Close()
}
```

After that, you'd just have to check the response code (should be self-explanatory), and the content-type (may need a comment), something like;

```go
if res.StatusCode != http.StatusOK {
	return false
}
if res.Header.Get("Content-Type") != "application/x-git-upload-pack-advertisement" {
	// not a smart server
	return false
}
return true
```

I dropped the `contentTypeResponse` variable, in favour of a comment explaining we detected it's not a "smart" server. The function description itself already explained that that means "shallow clone" is not supported

Although this would've worked for the return;

```go
return res.Header.Get("Content-Type") == "application/x-git-upload-pack-advertisement"
```

(mainly "taste"); I think it's less easy to grasp here, and having all returns `false` (early fail) makes it slightly more consistent